torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -0,0 +1,111 @@
+import math
+
+import torch
+import torch.distributed as dist
+
+
+def _orthogonalize(matrix, epsilon=1e-8):
+    """Applies Gram-Schmidt procedure to orthogonalize a given 2D tensor."""
+    num_cols = matrix.shape[1]
+    for i in range(num_cols):
+        # Normalize the i'th column.
+        col = matrix[:, i : i + 1]
+        # If no epsilon is added here, division by zero may be caused by vanishing gradients.
+        # This epsilon is not needed if the input matrix covers the gradients of at least one entire layer in the neural network.
+        if epsilon == 0:
+            # Note that col ** 2 can underflow/overflow if we use FP16.
+            # May need to consder multiplying a scaling factor and divding it later, or using bfloat16 isntead.
+            col /= torch.sqrt(torch.sum(col ** 2))
+        else:
+            col /= torch.sqrt(torch.sum(col ** 2)) + epsilon
+        # Project it on the rest and remove it.
+        if i + 1 < num_cols:
+            rest = matrix[:, i + 1 :]
+            rest -= torch.sum(col * rest, dim=0) * col
+
+
+def powerSGD_hook(
+    process_group: object, bucket: dist._GradBucket, rank: int = 1
+) -> torch.futures.Future:
+    """
+    This DDP communication hook implements a simplified PowerSGD gradient compression
+    algorithm described in https://arxiv.org/abs/1905.13727.
+    Once gradient tensors are aggregated across all workers, this hook applies
+    compression as follows:
+    1) Views the input flattened 1D gradient tensor as a square-shaped tensor M with 0 paddings;
+    2) Decomposes M into two low-rank tensors P and Q,
+    such that M = PQ^T, where Q is initialized from a standard normal distribution and orthogonalized;
+    2) Allreduces P;
+    3) Orthogonizes P;
```

Comment:
nit: it's "orthogonalizes" right?

Reply:
It's orthogonalized. See https://github.com/pytorch/pytorch/pull/48060#discussion_r525092833